core/lib/tasks/states.rake

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+def get_carmen_states(iso)
+  carmen_country = Carmen::Country.coded(iso)
+  return false unless carmen_country&.subregions?
+
+  if %w(IT).include? iso
+    return carmen_country.subregions.flat_map(&:subregions)
+  else
+    return carmen_country.subregions
+  end
+end
+
+def user_agrees(destroy_states, create_states, country)
+  puts "\e[31mStates for #{country.name} are incorrect. If you proceed, the following changes will happen:\e[0m"
+  puts "\e[31mThese states will be destroyed:\e[0m #{destroy_states}"
+  puts "\e[31mThese states will be created:\e[0m #{create_states}"
+  puts "\e[31mDo you want to proceed? [y/n]\e[0m "
+  ["y", "yes"].include? STDIN.gets.strip.downcase
+end
+
+def delete_states(country, carmen_names)
+  country.states.where.not(name: carmen_names).destroy_all
+end
+
+def create_states(country, carmen_states, state_names)
+  carmen_states.each do |state|
+    unless state_names.include? state.name
+      Spree::State.where(abbr: state.code, country: country).first_or_create!(
+        name: state.name
+      )
+    end
+  end
+end
+
+namespace :states do
+  desc "Regenerates states in cases where previous Solidus verions generated incorrect states"
+  task regenerate: :environment do
+    Spree::Country.all.each do |country|
+      if carmen_states = get_carmen_states(country.iso)
+        if (state_names = country.states.pluck(:name)) != (carmen_names = carmen_states.map(&:name).sort)
```

Comment:
Should we sort `country.states.pluck(:name)` as well in order to make sure that the ordering on this side is always correct?

Reply:
`country.states` is sorted through the association, do I don't think this would be an issue, but let me know if I'm wrong about that! https://github.com/solidusio/solidus/blob/7c76993eaafd6224f5c2e98fff5e8a33c2fd71b3/core/app/models/spree/country.rb#L5